crates/stages/src/stages/merkle.rs

```diff
@@ -35,33 +37,81 @@ pub const MERKLE_UNWIND: StageId = StageId("MerkleUnwind");
 #[derive(Debug)]
 pub enum MerkleStage {
     /// The execution portion of the hashing stage.
-    Execution,
+    Execution {
+        /// The threshold for switching from incremental trie building
+        /// of changes to whole rebuild. Num of transitions.
+        clean_threshold: u64,
+    },
     /// The unwind portion of the hasing stage.
     Unwind,
+
+    #[cfg(test)]
+    Both { clean_threshold: u64 },
 }
 
 #[async_trait::async_trait]
 impl<DB: Database> Stage<DB> for MerkleStage {
     /// Return the id of the stage
     fn id(&self) -> StageId {
         match self {
-            MerkleStage::Execution => MERKLE_EXECUTION,
+            MerkleStage::Execution { .. } => MERKLE_EXECUTION,
             MerkleStage::Unwind => MERKLE_UNWIND,
+            #[cfg(test)]
+            MerkleStage::Both { .. } => unreachable!(),
         }
     }
 
     /// Execute the stage.
     async fn execute(
         &mut self,
-        _tx: &mut Transaction<'_, DB>,
+        tx: &mut Transaction<'_, DB>,
         input: ExecInput,
     ) -> Result<ExecOutput, StageError> {
-        if matches!(self, MerkleStage::Unwind) {
-            info!(target: "sync::stages::merkle::unwind", "Stage is always skipped");
-            return Ok(ExecOutput { stage_progress: input.previous_stage_progress(), done: true })
-        }
+        let threshold = match self {
+            MerkleStage::Unwind => {
+                info!(target: "sync::stages::merkle::unwind", "Stage is always skipped");
+                return Ok(ExecOutput {
+                    stage_progress: input.previous_stage_progress(),
+                    done: true,
+                })
+            }
+            MerkleStage::Execution { clean_threshold } => *clean_threshold,
+            #[cfg(test)]
+            MerkleStage::Both { clean_threshold } => *clean_threshold,
```

Comment:
The test macros we use assume that the same stage is capable of both executing and unwinding. `Both` is used to allow this behaviour in tests. Although it could be removed with some modifications to the stage test suite.